pkg/plugins/file/main_test.go

```diff
@@ -91,6 +92,16 @@ func Test_Validate(t *testing.T) {
 					Exists: tt.mockFileExist,
 				},
 			}
+			file.files = make(map[string]string)
+			if len(file.spec.File) > 0 {
+				file.files[strings.TrimPrefix(file.spec.File, "file://")] = ""
+			}
+			// files
+			for _, currentFile := range file.spec.Files {
+				// TODO:? warn if already in? (duplicates)
+				// TODO:! only add if not already in
+				file.files[strings.TrimPrefix(currentFile, "file://")] = ""
+			}
```

Comment:
you should edit the existing list of cases, or add new ones.

I don't see any reason to change the test logic?

Reply:
(Must be doing it wrong) It doesn't go into main.New(), and thus the files map is empty.

I also don't know why the mock returns Exists: false

In a nut shell, will need help ^^

Reply:
no problem. We can have a session later today, or tomorrow if you want? (contact me in 1:1)

Reply:
Oh I see, you were in the correct direction, my aplogizie I've read too quickly (did not unfold the unmodified code the first time).

WDYT about changing the method validate to have a `file.Spec` receiver instead of a `file.File` ?
That would drastically simplifies the unit test logic here, as the SUT would be the provided spec instead of File?

Reply:
I've realized I didn't needed the file.files for main_test.validate, it passes without it.

But I'm still struggling with main_test.TestFile_Read() where I needed to keep f.files, and source_test.TestFile_Source()
I don't see how to call main.New() (should I?) in order to get File.files initialized.

Reply:
Validating the Spec instead of the File now, better.

Reply:
Good work, you're in the correct direction!

> But I'm still struggling with main_test.TestFile_Read() where I still need to initialize f.files, and source_test.TestFile_Source() (same kind of problem)

The idea of UT is to test an atomic piece of code without being influenced by the rest. The question (escept of the value of the test itself) in testing is always: what is the "System Under Test (SUT)", eg. what is the scope to be tested?

The unit tests in this case had been designed with the following rule in mind: Testing per public method of object (e.g. we don't test a full object, we don't test a package and we don't test private methods).

If you want to test a method such as `file.Read()`, then you have to build you own File{} object and then call the method on it. It allows testing not only inputs and outputs on the method only, but it also allows you to "tune" the object as you want (including mock injection instead of the default objects from the constructor) AND eventually checking the object (and mocks) state after calling the method to assert on mutations.


If you look at the existing tests, there should be the same pattern:
- Create a <whatever> object from the current test members (`tt.<something>` usually)
- Call the method to be tested on this object
- Assert on the results

Reply:
> You might want to provide the prebuilt File{} object with it internal files map to the method to be tested.

Doh! Thanks for the explainations 👍 
Will push a fix.

Reply:
Tried my best, because I find it hard to write down these things (compared to a synchronous communication) so it might be incomplete or perfectible, but at least it is a start: hope it helps.

If it does not help, don't hesitate to reach me for a 1:1 virtual coffee  :)

Reply:
Something like fca52ff?

Reply:
Exactly ! The declaration syntax of a map is a bit heavy (could declare a custom private type if it is really annoying) but otherwise that's perfect

Reply:
I'm not sure to understand the goal: the mock is designed to manipulate a single file and I don't see a reason (at least at first glance) to change it, since the code of the `file` package is changed to manipulate the file iterations.

The "mock" is there to allow isolate the interactions of the method being tested with the "rest of the system".

Reply:
The current mock return data for only one file, whereas I need it to be able to return the expected content of multiple files.

Reply:
I realize I don't want the mock to return info about multiple files, but I also need to specify the expected data of multiple files in target tests, that's why I've been looking in this direction.

Reply:
> I realize I don't want the mock to return info about multiple files, but I also need to specify the expected data of multiple files in target tests, that's why I've been looking in this direction.

I don't understand what do you mean?

Reply:
See here for example: https://github.com/updatecli/updatecli/pull/490/files?file-filters%5B%5D=.go&show-deleted-files=true&show-viewed-files=true#diff-dab848426c72c878db3ec93c1fb1650241e6a9d1a300d928c8decb0d2c01c6eaR78-R99
I need to be able to mock multiple expected file data

Reply:
I think I found how: `wantMockState       map[string]text.MockTextRetriever`

Reply:
Mmmh the mock might need to be updated to allow providing a custom function that would return the content, a bit like it's done with the http mock: 
- Mock definition: (The DoFunc is public and provided by caller): https://github.com/updatecli/updatecli/blob/main/pkg/core/httpclient/mocks.go
- Example fo usage: https://github.com/updatecli/updatecli/blob/f6981b693219a756a861cdb843bb047c5b372c5d/pkg/plugins/docker/dockerregistry/main_test.go#L174-L210
  - The test provides the function with the logic "if this URL then return A, else return B".

=> you migh want to update the text mock to provide such behavior, so the function would have the list of files and would return the value associated to key. WDYT?

Reply:
So... Wrong variables probably (should be mock*?), but something like that: ead5407

Reply:
I might be misunderstanding something but, there is only 1 `text.textRetriever` on a given `File{}`: how would you have the collection of mockS (it's a map of string -> MockRetreiever{} that you defined)?

Reply:
I think I need a collection of text.textRetriever, one for each file path of `files`
(Don't pay too much attention to the code of the last 2 commits, was more to get a feel of the kind of struct needed)

Reply:
I got a proposal which is a compromise:

- Add the "function" for the `MockRetriever` and add a logic like "if function == nil then {use <old behavior of the mock>}"
- On your test suite, add a `mockFile map[string]string` attribute on the test case struct, and fill this (close to what you did If i'm not mistaken)
- Define the function to be used for the mock one time in `file/main_test.go` as a private function.
- In the test's for loop, pass the packae private function to the mock

=> with this, you don't break the actual mock contract (if it is used somewhere else on the code), and you don't have to repeat the function (as I suppose you will need this function for source tests, conditions tests and target tests.. at least.

WDYT?